### DISCUSSION TITLE
Converting screen coordinates into lat/lon with a 2D plate-carre map and infinite scroll

### DISCUSSION THREAD

**Discussion by clinically-proven** [pelicanmapping/osgearth#2601] [Support Q&A] [answered]
Hi, I'm using code similar to the example here to show a 2D map that wraps around as the user pans east and west.

https://github.com/gwaldron/osgearth/blob/master/src/applications/osgearth_infinitescroll/osgearth_infinitescroll.cpp

It's working very well, but I'm having trouble converting client screen coordinates in the view to lat/lon coordinates. This is an overview of my conversion code, and it works correctly when the map is shown in 3D mode. In 2D infinite scroll mode, screenToWorld is returning true, dLat is either far out of range or 0, and dLon is 0. I'm guessing that something about the clever way the infinite scroll uses three matrices requires a more complex conversion. Any suggestions would be greatly appreciated. If it makes a difference, I'm using osgearth 2.8 on Windows.

	Viewer* m_viewer;
	EarthManipulator* m_earthMan;
	CPoint clientPoint;
	double dLat = 0;
	double dLon = 0;
	...
	// Convert MFC y value to OSG.
	clientPoint.y = m_viewer->getCamera()->getViewport()->height() - clientPoint.y;
	
	bool isValid = m_earthMan->screenToWorld(clientPoint.x, clientPoint.y, m_viewer, coord);	
	if (isValid)
	{
		// Convert the coord to a geopoint and get the lat/lon.				
		GeoPoint geoPoint;
		geoPoint.fromWorld(m_mapNode->getMapSRS(), coord);		
		dLat = geoPoint.y();
		dLon = geoPoint.x();		
	} //if

**Comment by gwaldron** [comment]
From what I can see, in version 2.8 the screenToWorld() computation ignores any transforms above the terrain. I am pretty sure this was addressed in later versions (specifically 3.4):
https://github.com/gwaldron/osgearth/blob/master/src/osgEarth/Terrain.cpp#L185

Version 2.8 is really old so the amount of support we can give you is somewhat limited.

**Comment by clinically-proven** [comment]
Thanks for the quick response. Version 2.8 is definitely quite ancient, but that's the version integrated into a 3rd party mapping product that we use. I'll see if there's any way to upgrade it.

**Comment by gwaldron** [comment]
Upgrading to the 3.x series is non-trivial; if you decide to go that route there's a quick guide here:
https://docs.osgearth.org/en/latest/3.0_upgrade_guide.html

And of course, we are always here to answer questions :)

**Reply by clinically-proven** [reply]
Great, thank you. It turns out that the third party library we use will be getting a big upgrade soon, including a move to osgearth 3.5. After we upgrade I'll verify that the issue is resolved. Thanks again for your help.

**Reply by gwaldron** [reply]
Great. May I ask what 3rd party library you are using?

**Reply by clinically-proven** [reply]
https://www.envitia.com/for-developers/maplink-pro/

**Comment by clinically-proven** [comment]
Hello again. We finally upgraded to a more current version of osgEarth, 3.7 to be exact. Unfortunately, screenToWorld is still returning bogus lat/lon values when the map uses 2D plate-carre projection. One important thing to note is that the issue happens even without the infinite scroll code in place. Let me know if you've got any suggestions about how to resolve this. I'm happy to provide more details if needed.

**Comment by gwaldron** [comment, marked as answer]
Try calling this instead:
```c++
GeoPoint p = mapNode->getGeoPointUnderMouse(view, x, y);
```
You can then convert this to long/lat coordinates by calling
```c++
GeoPoint p_geo = p.transform(p.getSRS()->getGeographicSRS());
double lat = p_geo.y();
double lon = p_geo.x();
```

**Comment by clinically-proven** [comment]
Thanks for the lightning-fast response. I tried your suggestion but am seeing similar results. For example, clicking in the center of South America yields a lat/lon of -1162693.04246448, -6530615.27944905. Interestingly, p and p_geo have the same x/y values. Any more suggestions?

**Reply by gwaldron** [reply]
Those are not long/lat, they are plate carre coordinates (aka Equidistant Cylindrical) and they look correct for South America.

The screenshot is taken from `osgearth_imgui readymap_flat.earth` and it uses the same code to calculate the X, Y and the Lat, Long:

<img width="679" alt="image" src="https://github.com/user-attachments/assets/ac89442d-7c51-4ab2-9ccf-f734ab8add70" />

**Comment by clinically-proven** [comment]
Thanks, this is all making sense now. In my implementation of the code above, I foolishly used the wrong SRS. With getGeographicSRS() I'm seeing the correct lat/lon values now. Thanks a bunch, and sorry for the trouble.

**Comment by clinically-proven** [comment]
The above method works great until the 2D map is put into infinite scroll mode. In this case, clicks within the center extent before the map is panned/scrolled return correct values, but clicks on the left or right sides and after significant scrolling return invalid (0,0 and null SRS) coordinates. Is there a way to translate local coordinates based on the scroll offset, such that `getGeoPointUnderMouse()` will handle them normally?

**Comment by gwaldron** [comment]
Not sure. Sadly we don't have the resources to dig into this at the moment. Infinite-scroll is not really an "official" feature so you will probably need to do some debugging to work it out :)

BTW, here is another infinite scrolling technique that uses multiple views instead of transforms. Feel free to check it out - it may work better for your setup:
https://gist.github.com/gwaldron/bc7d57cbdd6b93c515aae026ab605384

**Comment by clinically-proven** [comment]
Thanks, I'll give that a whirl.
